api/src/main/java/org/apache/iceberg/Metrics.java

```diff
@@ -120,4 +124,90 @@ public Long recordCount() {
   public Map<Integer, ByteBuffer> upperBounds() {
     return upperBounds;
   }
+
+  /**
+   * Implemented the method to enable serialization of ByteBuffers.
+   * @param out The stream where to write
+   * @throws IOException On serialization error
+   */
+  private void writeObject(ObjectOutputStream out) throws IOException {
+    out.writeObject(rowCount);
+    out.writeObject(columnSizes);
+    out.writeObject(valueCounts);
+    out.writeObject(nullValueCounts);
+
+    writeByteBufferMap(out, lowerBounds);
+    writeByteBufferMap(out, upperBounds);
+  }
```

Comment:
Is this serialization format, e.g. `rowCount, columnSizes, valueCounts, nullValueCounts` and then the `lowerBounds` followed by `upperBounds` currently used anywhere else?

If so, perhaps there's a utility class that can be used?

If not, I personally think that this serialization format should be  documented somewhere. We might need to include the format version in it as well, like we do for other places (where the new v2 format specification will be what is used to support row level deletes).

Additionally, would it make sense to move this serialization logic into a utility class if there's nothing that currently meets this need?

For example, type conversions to get the lower bounds value from the ByteBuffer are currently done in `org.apache.iceberg.types.Conversion#fromByteBuffer` as mentioned elsehwere in this file. Here's a link to the code, perhaps something there might make this task easier / more standardized: https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/types/Conversions.java

And if there's not a utility class, would it make sense to add one (like how the actual conversion of `upperBounds` and `lowerBounds` from their ByteBuffers is deletegated to the `Conversions` class?

Reply:
Oh, actually, I spoke to soon. There is documentation on the serialization of types when storing them as lower bounds / upper bounds in the maps of manifest files, namely in the `data_files` struct:  https://iceberg.apache.org/spec/#manifests

There's also documentation on what single values can be stored in the upper bounds / lower bounds byte buffer maps and how they are serialized: https://iceberg.apache.org/spec/#appendix-d-single-value-serialization

Perhaps something from that documentation page can be used to make this serialization choice  inline with the existing stuff? Or at the least I would recommend that documentation be added, though I think that you might want to consider submitting this PR to the dev mailing list at dv@iceberg.apache.org, as I feel that likely some people will have an opinion about the introduction of an additional standard for serialization (especially given that it's part of the public facing API). You can see the dev mailing list archive here: https://www.mail-archive.com/dev@iceberg.apache.org/

Reply:
This isn't really a serialization format that needs to be specified or reusable (besides maybe `writeByteBufferMap`). Java serialization should only be used within a running job that has just one version of the Iceberg library, so what matters is that a test can write an object and read an identical one.

We don't want to go too far down the path of documenting things like this that we don't intend to support as a format.